src/dune_pkg/opam_solver.ml

```diff
@@ -251,13 +259,37 @@ let opam_commands_to_actions (commands : OpamTypes.command list) =
        | None ->
          Pform.(Macro { macro = Pkg; payload = Payload.of_args [ "var"; "_"; variable ] }))
   in
-  List.filter_map commands ~f:(fun (args, _filter_opt) ->
+  List.filter_map commands ~f:(fun ((args, _filter_opt) as command) ->
+    let interpolate_opam_variables s =
+      Dune_re.Seq.split_full OpamFilter.string_interp_regex s
+      |> Seq.map ~f:(function
+        | `Text text -> `Text text
+        | `Delim group ->
+          (match Dune_re.Group.get group 0 with
+           | "%%" -> `Text "%"
+           | interp
+             when String.starts_with ~prefix:"%{" interp
+                  && String.ends_with ~suffix:"}%" interp ->
+             let ident = String.sub ~pos:2 ~len:(String.length interp - 4) interp in
+             `Pform (pform_of_ident ident)
+           | other ->
+             User_error.raise
+               [ Pp.textf
+                   "Encountered malformed variable interpolation while processing \
+                    commands for package %s."
+                   (OpamPackage.to_string package)
+               ; Pp.text "The variable interpolation:"
+               ; Pp.text other
```

Comment:
Rather than printing the string, could you extract the full location of the misformed string and use `Loc.pp`? That will print an excerpt with an underline.

Reply:
The problem with that would be that the commands are read from opam files using `OpamFile.OPAM.read` which doesn't tie the metadata fields back to locations in the file and the opam repo code inside dune doesn't give access to raw opam files. We could extend it to give access to raw files and use regexes to find the part of the file that contains the command but I'd rather do that in a separate feature if it's something we want to support.